core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java

```diff
@@ -52,17 +52,43 @@
     private final DictionaryId dictionarySourceId;
     private final boolean mayHaveNull;
 
-    public DictionaryBlock(int positionCount, Block dictionary, int[] ids)
+    public static Block create(int positionCount, Block dictionary, int[] ids)
     {
-        this(0, positionCount, dictionary, ids, false, false, randomDictionaryId());
+        return createInternal(positionCount, dictionary, ids, randomDictionaryId());
     }
 
     /**
      * This should not only be used when creating a projection of another dictionary block.
      */
-    public DictionaryBlock(int positionCount, Block dictionary, int[] ids, DictionaryId dictionarySourceId)
+    public static Block createProjectedDictionaryBlock(int positionCount, Block dictionary, int[] ids, DictionaryId dictionarySourceId)
```

Comment:
This method is very similar to `DictionaryBlock#getPositions` when it unwraps dictionaries. Yet `getPositions` has more optimizations like taking compactness into account or evaluating `uniqueIds`.
These optimizations improve e.g. serialization.